File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/EnumTypeTest.scala
package dev.typr.foundationssc.docs.core

import dev.typr.foundationssc.*
import dev.typr.foundationssc.connect.*

/** Verification that plain Scala 3 enums work with ofEnum — no extends java.lang.Enum needed. */
@SuppressWarnings(Array("unused"))
object EnumTypeTest:
  // Plain Scala 3 enum — NO extends java.lang.Enum
  enum Status:
    case PENDING, ACTIVE, COMPLETED

  // Just pass .values — works for any Scala 3 enum
  val pgStatus: PgType[Status] = PgTypes.ofEnum("status", Status.values)
  val duckStatus: DuckDbType[Status] = DuckDbTypes.ofEnum("status", Status.values)
  val mariaStatus: MariaType[Status] = MariaTypes.ofEnum(Status.values)
  val sqlServerStatus: SqlServerType[Status] = SqlServerTypes.ofEnum(Status.values)
  val db2Status: Db2Type[Status] = Db2Types.ofEnum(Status.values)
  val oracleStatus: OracleType[Status] = OracleTypes.ofEnum(Status.values)

  // Runtime verification with DuckDB
  def main(args: Array[String]): Unit =
    val tx = ConnectionSource.of(DuckDbConfig.inMemory().build()).transactor()

    val setup =
      Fragment
        .of("CREATE TYPE status AS ENUM('PENDING', 'ACTIVE', 'COMPLETED')")
        .execute()
        .combine(Fragment.of("CREATE TABLE enum_test (id INTEGER, status status)").execute())
        .combine(Fragment.of("INSERT INTO enum_test VALUES (1, ").value(duckStatus, Status.ACTIVE).append(")").execute())
        .combine(Fragment.of("INSERT INTO enum_test VALUES (2, ").value(duckStatus, Status.PENDING).append(")").execute())
        .voided()

    val querySingle = Fragment.of("SELECT status FROM enum_test WHERE id = 1").queryExactlyOne(duckStatus)
    val queryAll = Fragment.of("SELECT status FROM enum_test ORDER BY id").queryAll(duckStatus)

    val (result, all) = setup
      .combine(querySingle)
      .combine(queryAll)
      .map { case ((_, result), all) => (result, all) }
      .transact(tx)

    assert(result == Status.ACTIVE, s"Expected ACTIVE, got $result")
    assert(all == List(Status.ACTIVE, Status.PENDING), s"Expected [ACTIVE, PENDING], got $all")

    println("Scala 3 enum test passed — no extends java.lang.Enum needed!")

File: foundations-jdbc-test/src/kotlin/dev/typr/foundationskt/EnumTypeTest.kt
package dev.typr.foundationskt

import dev.typr.foundationskt.connect.*
import org.junit.Assert.assertEquals
import org.junit.Test

class EnumTypeTest {

    enum class Status { PENDING, ACTIVE, COMPLETED }

    @Test
    fun reifiedDuckDbEnum() {
        val statusType = DuckDbTypes.ofEnum<Status>("status")
        val tx = ConnectionSource.of(DuckDbConfig.inMemory().build()).transactor()

        tx.transact { conn ->
            Fragment.of("CREATE TYPE status AS ENUM('PENDING', 'ACTIVE', 'COMPLETED')").execute().run(conn)
            Fragment.of("CREATE TABLE eval_enum_test (id INTEGER, status status)").execute().run(conn)
            Fragment.of("INSERT INTO eval_enum_test VALUES (1, ")
                .value(statusType, Status.ACTIVE)
                .append(")")
                .execute().run(conn)

            val result = Fragment.of("SELECT status FROM eval_enum_test WHERE id = 1")
                .queryExactlyOne(statusType)
                .run(conn)

            assertEquals(Status.ACTIVE, result)
        }
    }

    @Test
    fun reifiedMariaDbEnum() {
        val statusType = MariaTypes.ofEnum<Status>()
        val rendered = Fragment.of("SELECT ").value(statusType, Status.PENDING).render()
        assertEquals("SELECT ?", rendered)
    }

    @Test
    fun reifiedSqlServerEnum() {
        val statusType = SqlServerTypes.ofEnum<Status>()
        val rendered = Fragment.of("SELECT ").value(statusType, Status.COMPLETED).render()
        assertEquals("SELECT ?", rendered)
    }

    @Test
    fun reifiedDb2Enum() {
        val statusType = Db2Types.ofEnum<Status>()
        val rendered = Fragment.of("SELECT ").value(statusType, Status.ACTIVE).render()
        assertEquals("SELECT ?", rendered)
    }

    @Test
    fun valuesBasedDuckDbEnum() {
        val statusType = DuckDbTypes.ofEnum("status", Status.values())
        val tx = ConnectionSource.of(DuckDbConfig.inMemory().build()).transactor()

        tx.transact { conn ->
            Fragment.of("CREATE TYPE status AS ENUM('PENDING', 'ACTIVE', 'COMPLETED')").execute().run(conn)
            Fragment.of("CREATE TABLE eval_enum_test2 (id INTEGER, status status)").execute().run(conn)
            Fragment.of("INSERT INTO eval_enum_test2 VALUES (1, ")
                .value(statusType, Status.COMPLETED)
                .append(")")
                .execute().run(conn)

            val result = Fragment.of("SELECT status FROM eval_enum_test2 WHERE id = 1")
                .queryExactlyOne(statusType)
                .run(conn)

            assertEquals(Status.COMPLETED, result)
        }
    }
}
